File: sample/experiments-compiler/experiment-enums-compiler/java/src/main/java/com/uber/crumb/sample/experimentenumscompiler/ExperimentsCompiler.java
/*
 * Copyright 2018. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.uber.crumb.sample.experimentenumscompiler;

import static com.google.auto.common.MoreElements.isAnnotationPresent;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toMap;
import static java.util.stream.Collectors.toSet;
import static javax.lang.model.element.Modifier.ABSTRACT;
import static javax.lang.model.element.Modifier.FINAL;
import static javax.lang.model.element.Modifier.PUBLIC;
import static javax.lang.model.element.Modifier.STATIC;

import com.google.auto.common.MoreElements;
import com.google.auto.service.AutoService;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.FieldSpec;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;
import com.uber.crumb.compiler.api.CrumbConsumerExtension;
import com.uber.crumb.compiler.api.CrumbContext;
import com.uber.crumb.compiler.api.CrumbProducerExtension;
import com.uber.crumb.sample.experimentsenumscompiler.annotations.Experiments;
import com.uber.crumb.sample.experimentsenumscompiler.annotations.ExperimentsCollector;
import java.io.IOException;
import java.lang.annotation.Annotation;
import java.util.Arrays;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.TypeElement;
import javax.tools.Diagnostic;

/**
 * A simple crumb producer/consumer that reads "experiment enums" from libraries and writes a
 * mapping record of them in a consuming app. Could be useful for tracking what experiments are on
 * the classpath.
 */
@AutoService({CrumbProducerExtension.class, CrumbConsumerExtension.class})
public final class ExperimentsCompiler implements CrumbProducerExtension, CrumbConsumerExtension {

  private static final String METADATA_KEY = "ExperimentsCompiler";

  @Override
  public boolean isConsumerApplicable(
      CrumbContext context, TypeElement type, Collection<? extends AnnotationMirror> annotations) {
    return isAnnotationPresent(type, ExperimentsCollector.class);
  }

  @Override
  public boolean isProducerApplicable(
      CrumbContext context, TypeElement type, Collection<? extends AnnotationMirror> annotations) {
    return isAnnotationPresent(type, Experiments.class);
  }

  @Override
  public void consume(
      CrumbContext context,
      TypeElement type,
      Collection<? extends AnnotationMirror> annotations,
      Set<? extends Map<String, String>> metadata) {

    // Must be an abstract class because we're generating the backing implementation.
    if (type.getKind() != ElementKind.CLASS) {
      context
          .getProcessingEnv()
          .getMessager()
          .printMessage(
              Diagnostic.Kind.ERROR,
              "@"
                  + ExperimentsCollector.class.getSimpleName()
                  + " is only applicable on classes when consuming!",
              type);
      return;
    } else if (!type.getModifiers().contains(ABSTRACT)) {
      context
          .getProcessingEnv()
          .getMessager()
          .printMessage(Diagnostic.Kind.ERROR, "Must be abstract!", type);
    }

    // Map of enum TypeElement to its members
    Map<TypeElement, Set<String>> experimentClasses =
        metadata
            .stream()
            .map(data -> data.get(METADATA_KEY))
            .map(
                enumClass -> context.getProcessingEnv().getElementUtils().getTypeElement(enumClass))
            .collect(
                toMap(
                    typeElement -> typeElement,
                    typeElement ->
                        typeElement
                            .getEnclosedElements()
                            .stream()
                            .filter(e -> e.getKind() == ElementKind.ENUM_CONSTANT)
                            .map(Object::toString)
                            .collect(toSet())));

    FieldSpec mapField =
        FieldSpec.builder(
                ParameterizedTypeName.get(
                    ClassName.get(Map.class),
                    TypeName.get(Class.class),
                    ParameterizedTypeName.get(List.class, String.class)),
                "EXPERIMENTS",
                PUBLIC,
                STATIC,
                FINAL)
            .initializer(CodeBlock.of("new $T<>()", LinkedHashMap.class))
            .build();

    CodeBlock.Builder staticInitBlock = CodeBlock.builder();
    experimentClasses.forEach(
        (key, value) ->
            staticInitBlock.addStatement(
                "$N.put($T.class, $T.asList($L))",
                mapField,
                TypeName.get(key.asType()),
                TypeName.get(Arrays.class),
                String.join(", ", value.stream().map(s -> "\"" + s + "\"").collect(toList()))));

    TypeSpec generatedClass =
        TypeSpec.classBuilder("Experiments_" + type.getSimpleName().toString())
            .addModifiers(PUBLIC, FINAL)
            .superclass(TypeName.get(type.asType()))
            .addField(mapField)
            .addStaticBlock(staticInitBlock.build())
            .build();

    try {
      JavaFile.builder(MoreElements.getPackage(type).getQualifiedName().toString(), generatedClass)
          .build()
          .writeTo(context.getProcessingEnv().getFiler());
    } catch (IOException e) {
      context
          .getProcessingEnv()
          .getMessager()
          .printMessage(
              Diagnostic.Kind.ERROR,
              "Failed to write generated experiments mapping! " + e.getMessage(),
              type);
    }
  }

  @Override
  public Map<String, String> produce(
      CrumbContext context, TypeElement type, Collection<? extends AnnotationMirror> annotations) {
    if (type.getKind() != ElementKind.ENUM) {
      context
          .getProcessingEnv()
          .getMessager()
          .printMessage(
              Diagnostic.Kind.ERROR,
              "@"
                  + Experiments.class.getSimpleName()
                  + " is only applicable on enums when producing!",
              type);
      return ImmutableMap.of();
    }
    return ImmutableMap.of(METADATA_KEY, type.getQualifiedName().toString());
  }

  @Override
  public Set<Class<? extends Annotation>> supportedConsumerAnnotations() {
    return ImmutableSet.of(ExperimentsCollector.class);
  }

  @Override
  public Set<Class<? extends Annotation>> supportedProducerAnnotations() {
    return ImmutableSet.of(Experiments.class);
  }

  @Override
  public String key() {
    return METADATA_KEY;
  }
}

File: sample/experiments-compiler/experiment-enums-compiler/kotlin/src/main/kotlin/com/uber/crumb/sample/experimentenumscompiler/ExperimentsCompiler.kt
/*
 * Copyright 2020. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.uber.crumb.sample.experimentenumscompiler

import com.google.auto.common.MoreElements.isAnnotationPresent
import com.google.auto.service.AutoService
import com.squareup.kotlinpoet.FileSpec
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.STAR
import com.squareup.kotlinpoet.WildcardTypeName
import com.squareup.kotlinpoet.asClassName
import com.squareup.kotlinpoet.asTypeName
import com.squareup.kotlinpoet.classinspector.elements.ElementsClassInspector
import com.squareup.kotlinpoet.metadata.isObject
import com.squareup.kotlinpoet.metadata.specs.toFileSpec
import com.squareup.kotlinpoet.metadata.toImmutableKmClass
import com.uber.crumb.compiler.api.ConsumerMetadata
import com.uber.crumb.compiler.api.CrumbConsumerExtension
import com.uber.crumb.compiler.api.CrumbContext
import com.uber.crumb.compiler.api.CrumbExtension.IncrementalExtensionType
import com.uber.crumb.compiler.api.CrumbExtension.IncrementalExtensionType.ISOLATING
import com.uber.crumb.compiler.api.CrumbProducerExtension
import com.uber.crumb.compiler.api.ExtensionKey
import com.uber.crumb.compiler.api.ProducerMetadata
import com.uber.crumb.sample.experimentsenumscompiler.annotations.Experiments
import com.uber.crumb.sample.experimentsenumscompiler.annotations.ExperimentsCollector
import javax.annotation.processing.ProcessingEnvironment
import javax.lang.model.element.AnnotationMirror
import javax.lang.model.element.Element
import javax.lang.model.element.ElementKind.CLASS
import javax.lang.model.element.ElementKind.ENUM
import javax.lang.model.element.ElementKind.ENUM_CONSTANT
import javax.lang.model.element.TypeElement
import javax.tools.Diagnostic.Kind.ERROR

/**
 * A simple crumb producer/consumer that reads "experiment enums" from libraries and writes a
 * mapping record of them in a consuming app. Could be useful for tracking what experiments are on
 * the classpath.
 */
@AutoService(CrumbProducerExtension::class, CrumbConsumerExtension::class)
class ExperimentsCompiler : CrumbProducerExtension, CrumbConsumerExtension {

  companion object {
    private const val METADATA_KEY: ExtensionKey = "ExperimentsCompiler"
  }

  override fun isConsumerApplicable(
    context: CrumbContext,
    type: TypeElement,
    annotations: Collection<AnnotationMirror>
  ): Boolean {
    return isAnnotationPresent(type, ExperimentsCollector::class.java)
  }

  override fun isProducerApplicable(
    context: CrumbContext,
    type: TypeElement,
    annotations: Collection<AnnotationMirror>
  ): Boolean {
    return isAnnotationPresent(type, Experiments::class.java)
  }

  /** This is isolating because it only depends on the consumer type instance.  */
  override fun consumerIncrementalType(
    processingEnvironment: ProcessingEnvironment
  ): IncrementalExtensionType = ISOLATING

  override fun consume(
    context: CrumbContext,
    type: TypeElement,
    annotations: Collection<AnnotationMirror>,
    metadata: Set<ConsumerMetadata>
  ) {

    // Must be a type that supports extension values
    if (type.kind != CLASS) {
      context.processingEnv
        .messager
        .printMessage(
          ERROR,
          "@${ExperimentsCollector::class.java.simpleName} is only applicable on classes when consuming!",
          type
        )
      return
    }

    val kmClass = type.toImmutableKmClass()

    // Must be an object
    if (!kmClass.isObject) {
      context.processingEnv
        .messager
        .printMessage(
          ERROR,
          "@${ExperimentsCollector::class.java.simpleName} can't be applied to $type: must be a Kotlin object class",
          type
        )
      return
    }

    val elementsInspector = ElementsClassInspector.create(context.processingEnv.elementUtils, context.processingEnv.typeUtils)
    val spec = kmClass.toFileSpec(elementsInspector, type.asClassName())
    val packageName = spec.packageName

    // Map of enum TypeElement to its members
    val experimentClasses = metadata
      .mapNotNull { it[METADATA_KEY] }
      .map { enumClass -> context.processingEnv.elementUtils.getTypeElement(enumClass) }
      .associateWith {
        it.enclosedElements
          .filter { it.kind == ENUM_CONSTANT }
          .map(Element::toString)
      }

    val initializerCode = experimentClasses
      .map { "%T::class.java to listOf(${it.value.joinToString(", ") { "%S" }})" }
      .joinToString()
    val initializerValues = experimentClasses
      .flatMap { listOf(it.key.asClassName(), *it.value.toTypedArray()) }
      .toTypedArray()
    val mapFunction = FunSpec.builder("experiments")
      .receiver(type.asClassName())
      .returns(
        Map::class.asClassName().parameterizedBy(
          Class::class.asClassName().parameterizedBy(
            WildcardTypeName.producerOf(
              Enum::class.asClassName().parameterizedBy(STAR)
            )
          ),
          List::class.asClassName().parameterizedBy(
            String::class.asTypeName()
          )
        )
      )
      .addStatement("return mapOf($initializerCode)", *initializerValues)
      .addOriginatingElement(type)
      .build()

    // Generate the file
    FileSpec.builder(packageName, "${type.simpleName}_Experiments")
      .addFunction(mapFunction)
      .build()
      .writeTo(context.processingEnv.filer)
  }

  override fun producerIncrementalType(
    processingEnvironment: ProcessingEnvironment
  ): IncrementalExtensionType = ISOLATING

  override fun produce(
    context: CrumbContext,
    type: TypeElement,
    annotations: Collection<AnnotationMirror>
  ): ProducerMetadata {
    if (type.kind != ENUM) {
      context.processingEnv
        .messager
        .printMessage(
          ERROR,
          "@${Experiments::class.java.simpleName} is only applicable on enums when producing!",
          type
        )
      return emptyMap<String, String>() to emptySet()
    }
    return mapOf(METADATA_KEY to type.qualifiedName.toString()) to setOf(type)
  }

  override fun supportedConsumerAnnotations(): Set<Class<out Annotation>> = setOf(
    ExperimentsCollector::class.java
  )

  override fun supportedProducerAnnotations(): Set<Class<out Annotation>> = setOf(
    Experiments::class.java
  )

  override val key = METADATA_KEY
}
